### PR TITLE
Check that all needs are met and close alert

### LINKED ISSUE #154: closed alerts don't send final txt msg
From Megan:  I am registered as a Provider in the RAN for some of the services.  I received an alert from Donna Butera on Friday via text message.  When I logged on to the RAN it was there as well but when she closed the case, I did not get a text message. I had to log back in to see that it was closed.  Something for the Devs to check out.

### PR DESCRIPTION
**Pull request by justinnoah**
fixes #154

- Moves heavily lifting of need management into a celery task.
- Resolves a need as normal and adds a resolution check for all needs on an alert
- Sends out a final message to providers_notified via email and text
- Minor code cleanup as well

### CODE REVIEW

File: _15thnight/api/need.py [hunk 1]
```diff
@@ -24,22 +23,27 @@ def get_need(need_id):
...
-    need.resolve_message = form.message.data
-    for provision in need.provisions:
-        provision.selected = provision.id in form.provisions.data
-    send_out_resolution(need)
-    need.save()
-    return '', 200
\ No newline at end of file
+
+    # Check if alert is resolved, notify those involved
+    queue_resolve_alert.apply_async()
```
Comment (nobane) [inline _15thnight/api/need.py]: The pattern the project is following a pattern of calling functions from `core`. `core` does asynchronous things. Please leave this as `send_out_resolution`.

File: _15thnight/models.py [hunk 2]
```diff
@@ -194,7 +198,7 @@ def get_alerts(cls):
     def get_provider_alerts(cls, provider):
         alerts = Alert.query \
             .join(cls.providers_notified) \
-            .filter(ProviderNotified.provider_id==provider.id) \
+            .filter(ProviderNotified.provider_id == provider.id) \
```
Comment (nobane) [inline _15thnight/models.py:205]: Inside filter statements, we do not space out the equality check. Please change these back.

**Comment by nobane** [comment]
"Heavy lifting" is done in `core`. Asynchronous functions (celery tasks) should be as light as possible. As a rule, our API endpoints do not trigger asynchronous tasks. They can call methods from core. Those methods in core trigger asynchronous actions.

**Comment by nobane** [comment]
There is a circular import here.
`core` is importing from `queue` and `queue` imports from `core`.

As a rule:
`api` can _only_ import from `core` and can _never_ import from `queue`
`core` _always_ imports from `queue`
`queue` never imports from `core` and generally imports as little as possible.

**Comment by nobane** [comment]
We do not open PRs for tracking purposes. PRs are for finished, tested and complete work. The PR indicates this is a fix, but it is clearly untested and a PoC.

File: CHANGELOG [hunk 3]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 
 ## [Unreleased]
+### Added
+- is_open method on Alert model
+- Send resolution notice(s) if not Alert.is_open()
```
Comment (nobane) [inline CHANGELOG]: Changelog doesn't need to get this technical. I want it to be more for what effects our end users will notice. So here I would just say `Send our resolution notices if all needs have been marked resolved`.

File: _15thnight/models.py (hunk 2 above)
Comment (justinnoah) [reply _15thnight/models.py:205]: I thought our coding style for this project was to follow standard pep8. Removal of those white spaces causes pep8 to through an E225.

http://pep8.readthedocs.io/en/latest/intro.html#error-codes

File: _15thnight/api/account.py [hunk 4]
```diff
@@ -74,16 +74,17 @@ def forgot_password():
                 user.reset_created_at < datetime.utcnow() - reset_token_life):
             user.generate_reset_token()
         user.save()
+
         link = '%sreset-password/%s/%s' % (
-            url_for('index', _external=True), user.email, user.reset_token
-        )
+            url_for('index', _external=True), user.email, user.reset_token)
```
Comment (nobane) [inline _15thnight/api/account.py]: Let's keep this formatting the way it was originally. We are doing it below for `Message`. Consistency is important.

File: _15thnight/queue.py [hunk 5]
```diff
@@ -35,10 +34,25 @@ def queue_send_message(email, number, subject, body):
...
+    if not need.alert.is_open():
+        alert = need.alert
+        body = "Alert {0} has been closed. See {1} for details" % (
+            alert, alert.url())
+        providers = [provider for provider in alert.providers_notified]
+        subject = "Alert %s has closed." % alert.id
+
+        for provider in providers:
+            queue_send_message(
+                provider.email, provider.phone_number, subject, body)
```
Comment (nobane) [inline _15thnight/queue.py]: To avoid complexity, `queue` tasks should not be firing off other `queue` tasks. Functions in `core` fire off all queue tasks.

File: _15thnight/api/need.py [hunk 6]
```diff
@@ -24,22 +25,36 @@ def get_need(need_id):
...
     need.resolve_message = form.message.data
     for provision in need.provisions:
         provision.selected = provision.id in form.provisions.data
-    send_out_resolution(need)
     need.save()
-    return '', 200
\ No newline at end of file
+
+    # Check if alert is resolved, notify those involved
+    resolve_alert
```
Comment (nobane) [inline _15thnight/api/need.py]: You're not calling the function here?

File: _15thnight/api/account.py (hunk 4 above)
Comment (justinnoah) [reply _15thnight/api/account.py]: Many times you have mentioned (specifically with imports anyway) that if all the code, including final paren, can fit on one line, it should. That is why I changed this.

File: _15thnight/api/need.py (hunk 6 above)
Comment (nobane) [reply _15thnight/api/need.py]: Also we don't know if we are resolving just a need or the whole alert. This should still call a function called `send_out_resolution`

Comment (justinnoah) [reply _15thnight/api/need.py]: oops, yeah

File: _15thnight/queue.py (hunk 5 above)
Comment (nobane) [reply _15thnight/queue.py]: This logic should be moved to the `send_out_resolution` function in `core` and will then call `queue_send_message`.

**Review by nobane — DISMISSED** [review]

File: _15thnight/queue.py (hunk 5 above)
Comment (justinnoah) [reply _15thnight/queue.py]: This is in queue because alert.is_open has a potential to cause too long of a wait for the end user. And this only adds one function call which is trivial. Since apply_async is **not** being called on the task we are not starting more tasks from within this task.

**Comment by justinnoah** [comment]
@spectralsun conflicts resolved and questions/concerned responded to

File: _15thnight/queue.py [hunk 7]
```diff
@@ -35,10 +34,28 @@ def queue_send_message(email, number, subject, body):
...
+
+@celery.task
+def queue_send_alert_resolution_notice(need):
+    # Check if alert is closed and trigger an alert closed notification
+    if not need.alert.is_open():
+        alert = need.alert
+        body = (
+            "Alert sent at {0} for a {1}yo {2} with {3} needs has been closed."
+            "\nSee: {4}" % (
+                alert.created_at, alert.age, alert.gender, alert.need_count(),
```
Comment (nobane) [inline _15thnight/queue.py]: Gender could be unspecified, in that case we wouldn't insert it because the original alert would just say `Xy/o`

File: _15thnight/core.py [hunk 8]
```diff
@@ -95,10 +96,26 @@ def respond_to_alert(provider, needs_provided, alert):
...
-def send_out_resolution(need):
-    """
-    Sends out a resolution for a need to providers.
-    """
+def send_alert_resolution_notice(need):
+    # Check if alert is closed and trigger an alert closed notification
+    if not need.alert.is_open():
+        alert = need.alert
+        body = "Alert sent at {0} for a {1}yo " % (alert.created_at, alert.age)
+        body += "{0} " % alert.gender if alert.gender.lower() in [
```
Comment (justinnoah) [inline _15thnight/core.py]: @spectralsun i think this is what you were looking for instead?

File: _15thnight/core.py [hunk 9]
```diff
@@ -95,10 +96,26 @@ def respond_to_alert(provider, needs_provided, alert):
     return response
 
 
-def send_out_resolution(need):
-    """
-    Sends out a resolution for a need to providers.
-    """
+def send_alert_resolution_notice(need):
```
Comment (nobane) [inline _15thnight/core.py]: Move the code in this function into the bottom of `resolve_need`. No need for this to be in its own function.

File: _15thnight/core.py [hunk 10]
```diff
@@ -95,10 +96,26 @@ def respond_to_alert(provider, needs_provided, alert):
...
+        providers = [provider for provider in alert.providers_notified]
+        subject = "Alert %s has closed." % alert.id
+
+        for provider in providers:
+            queue_send_message.apply_async(args=[
+                provider.email, provider.phone_number, subject, body])
+
+
+def resolve_need(need):
     alert = need.alert
```
Comment (nobane) [inline _15thnight/core.py:106]: Put a multiline comment here that says: `Resolve a need and set alert as closed when all needs are resolved`

File: _15thnight/models.py [hunk 11]
```diff
@@ -1,10 +1,11 @@
 """Data Models."""
-import uuid
 from datetime import datetime, timedelta
+import uuid
```
Comment (nobane) [inline _15thnight/models.py:2]: Why not `import` statements before `from ... import ...` statements?

File: _15thnight/core.py [hunk 12]
```diff
@@ -130,3 +148,6 @@ def send_out_resolution(need):
                 body=body
             )
         )
+
+    # Check if alert is closed, if so, send out resolution notices
+    send_alert_resolution_notice.apply_async(need)
```
Comment (nobane) [inline _15thnight/core.py]: Will `apply_async` work when you call it this way?

File: _15thnight/core.py (hunk 8 above)
Comment (nobane) [reply _15thnight/core.py]: Close, can be done simpler with:
`gender = '' if alert.gender == 'unspecified' else alert.gender`

Comment (nobane) [reply _15thnight/core.py]: How will `created_at` look for this message?

File: _15thnight/models.py (hunk 11 above)
Comment (justinnoah) [reply _15thnight/models.py:2]: I typically sort by module name being imported. In this case, uuid comes after datetime alphabetically. I also group, in order, stdlib imports, 3rd party imports, project imports.

File: _15thnight/core.py (hunk 12 above)
Comment (justinnoah) [reply _15thnight/core.py]: nice catch, i do not believe so.

File: _15thnight/core.py (hunk 9 above)
Comment (justinnoah) [reply _15thnight/core.py]: I feel like the resolve_need method is starting to get too large, hence the code breakup. This code breakup also allows for easier testing and an easier path for a possible future "resend final alert closed notification" if needed.

File: _15thnight/models.py (hunk 11 above)
Comment (nobane) [reply _15thnight/models.py:2]: I agree with the grouping and that is what is used through the projects.

But within the subgroups I list `import ...` first before `from ... import ...`

File: _15thnight/api/need.py [hunk 13]
```diff
@@ -23,23 +24,37 @@ def get_need(need_id):
...
+
+    # Check validity of need_id
     if not need:
         return api_error('Need not found')
     if current_user.id != need.alert.user_id:
         return api_error('Permission denied')
     if need.resolved:
         return api_error('Need already resolved!')
+
+    # valid need_id given, validate the form
```
Comment (nobane) [inline _15thnight/api/need.py]: No need for this comment here.

File: _15thnight/api/need.py [hunk 14]
```diff
@@ -23,23 +24,37 @@ def get_need(need_id):
...
     form = ResolveNeedForm(need=need)
     if not form.validate_on_submit():
         return api_error(form.errors)
+
+    # Store form data as a the Need
```
Comment (nobane) [inline _15thnight/api/need.py]: This comment doesn't make sense. Just reword it to `Update need as resolved with notes and message`

File: _15thnight/api/need.py [hunk 15]
```diff
@@ -23,23 +24,37 @@ def get_need(need_id):
...
         provision.selected = provision.id in form.provisions.data
-    send_out_resolution(need)
     need.save()
-    return '', 200
\ No newline at end of file
+
+    # Check if alert is resolved, notify those involved
+    resolve_need(need)
+
+    # Return success
```
Comment (nobane) [inline _15thnight/api/need.py]: Remove this comment

File: _15thnight/core.py [hunk 16]
```diff
@@ -24,10 +25,10 @@ def send_out_alert(alert_form):
     providers = User.providers_with_services(need_ids)
     for provider in providers:
         needs_provided = [
-            need for need in provider.services if need.id in need_ids
+            need_ for need_ in provider.services if need_.id in need_ids
```
Comment (nobane) [inline _15thnight/core.py:30]: This rename is not needed because the previous `need` declaration is in a for loop.

File: _15thnight/core.py [hunk 17]
```diff
@@ -24,10 +25,10 @@ def send_out_alert(alert_form):
...
         ]
         gender = ' ' + alert.gender if alert.gender != 'unspecified' else ''
-        needs = ", ".join([need.name for need in needs_provided])
+        needs = ", ".join([need_.name for need_ in needs_provided])
```
Comment (nobane) [inline _15thnight/core.py]: Again, rename not needed.

File: _15thnight/core.py [hunk 18]
```diff
@@ -40,7 +41,7 @@ def send_out_alert(alert_form):
         provider_notified = ProviderNotified(
             provider=provider,
             alert=alert,
-            needs=Need.get_by_ids([need.id for need in needs_provided])
+            needs=Need.get_by_ids([need_.id for need_ in needs_provided])
```
Comment (nobane) [inline _15thnight/core.py]: Rename not needed.

File: _15thnight/core.py [hunk 19]
```diff
@@ -95,10 +96,9 @@ def respond_to_alert(provider, needs_provided, alert):
     return response
 
 
-def send_out_resolution(need):
-    """
-    Sends out a resolution for a need to providers.
-    """
+def resolve_need(need):
+    """Resolve a need and trigger an alert closed if necessary."""
```
Comment (nobane) [inline _15thnight/core.py]: Keep this as a multiline comment please.

File: _15thnight/core.py [hunk 20]
```diff
@@ -130,3 +131,26 @@ def send_out_resolution(need):
                 body=body
             )
         )
+
+    # Check if alert is closed, if so, send out resolution notices
+    send_alert_resolution_notice.apply_async(args=[need])
```
Comment (nobane) [inline _15thnight/core.py]: This doesn't need to be async. Delete lines 136-140 please. It can all be one function. `Alert.is_open` is not expensive and will work quickly. Things that need to be async are actions like communicating to a remote web or mail server. Communicating to a database (probably local) that the app has bound to performs fine.

File: _15thnight/core.py (hunk 16 above)
Comment (justinnoah) [reply _15thnight/core.py:30]: True, though need is still technically defined. The variable defined by the for loop is still accessible after the for loop exits.

Try the following in your python shell to see:
```python
for a in range(0,2):
  print(a)
print(a)
```
Output will be:
```
0
1
1
```

File: _15thnight/models.py (hunk 11 above)
Comment (justinnoah) [reply _15thnight/models.py:2]: Looks weird to me that way, but will do

File: _15thnight/core.py (hunk 17 above)
Comment (nobane) [reply _15thnight/core.py]: Rename to `need_provided`

File: _15thnight/core.py (hunk 20 above)
Comment (justinnoah) [reply _15thnight/core.py]: Oh, snap. This couldn't have been ran that way. send_alert_resolution_notice isn't a celery task. I'm still not a fan of such a large method and it needs to be separate for testing anyway.

File: _15thnight/core.py (hunk 18 above)
Comment (nobane) [reply _15thnight/core.py]: rename to `need_provided`

**Comment by justinnoah** [comment]
@spectralsun fixedup per comments and conflicts resolved

File: _15thnight/core.py [hunk 21]
```diff
@@ -133,6 +135,30 @@ def send_out_resolution(need):
...
+    _send_alert_resolution_notice(need)
+
+
+def _send_alert_resolution_notice(need):
+    """
+    Check if need.alert is closed and trigger an alert closed notification.
+    """
+    if not need.alert.is_open():
+        alert = need.alert
+        body = "Alert sent at {0} for a {1}yo " % (alert.created_at, alert.age)
```
Comment (nobane) [inline _15thnight/core.py]: How will `created_at` look in this text message?

Comment (nobane) [reply _15thnight/core.py]: Format it to `MM/DD hh:mm[am,pm]`

File: _15thnight/models.py [hunk 22]
```diff
@@ -246,16 +250,42 @@ def to_provider_json(self, provider):
...
+    def get_gender(self):
+        """
+        Return a formatted string with the gender if male/female.
+        Blank string if unspecified.
+        """
+        gender = '' if self.gender == 'unspecified' else self.gender
+        g_string = " %s" % gender
+        return g_string
+
+    def url(self):
```
Comment (nobane) [inline _15thnight/models.py:276]: In `core.py` you're referencing this as a property. Let's make `url` a property function.

File: _15thnight/models.py [hunk 23]
```diff
@@ -246,16 +250,42 @@ def to_provider_json(self, provider):
...
+        return True
+
+    def get_gender(self):
+        """
+        Return a formatted string with the gender if male/female.
+        Blank string if unspecified.
+        """
+        gender = '' if self.gender == 'unspecified' else self.gender
+        g_string = " %s" % gender
+        return g_string
```
Comment (nobane) [inline _15thnight/models.py]: While this *is* hilarious, please either rename this variable or refactor it to not need it.

File: _15thnight/core.py [hunk 24]
```diff
@@ -98,16 +99,17 @@ def respond_to_alert(provider, needs_provided, alert):
     return response
 
 
-def send_out_resolution(need):
+def resolve_need(need):
     """
-    Sends out a resolution for a need to providers.
+    Resolve a need and trigger an alert closed if necessary.
     """
+    print("resolving need: %s" % need.id)
```
Comment (nobane) [inline _15thnight/core.py]: Remove this print statement.

File: _15thnight/core.py [hunk 25]
```diff
@@ -133,6 +136,34 @@ def send_out_resolution(need):
                 body=body
             )
         )
+    print("wtf")
```
Comment (nobane) [inline _15thnight/core.py]: Remove this print statement.

File: _15thnight/core.py [hunk 26]
```diff
@@ -133,6 +136,34 @@ def send_out_resolution(need):
...
+        created = alert.created_at.strftime("%m/%d at %I:%M %p")
+        body = (
+            "Alert sent on %s for a %syo%s has been closed."
+            "\nSee%s"
+            % (created, age, gender, alert.url))
+        providers = [
+            provider_notified.provider for provider_notified
+            in alert.providers_notified
+        ]
+        print(providers)
```
Comment (nobane) [inline _15thnight/core.py]: Remove this print statement.

File: _15thnight/core.py [hunk 27]
```diff
@@ -133,6 +136,34 @@ def send_out_resolution(need):
...
+    Check if need.alert is closed and trigger an alert closed notification.
+    """
+    if need.alert.is_closed:
+        alert = need.alert
+        age = alert.age
+        gender = alert.get_gender()
+        created = alert.created_at.strftime("%m/%d at %I:%M %p")
+        body = (
+            "Alert sent on %s for a %syo%s has been closed."
+            "\nSee%s"
```
Comment (nobane) [inline _15thnight/core.py]: This line should be `"\nSee %s"` (add a space after `See`).

**Comment by justinnoah** [comment]
@spectralsun prints removed, spacing fixed. Good catch on those!
